From d167456c120d5790e16fadf5bfd13c2787eb6c20 Mon Sep 17 00:00:00 2001
From: Steve deRosier <Steve.Derosier@lairdtech.com>
Date: Mon, 8 Jul 2013 15:49:39 -0700
Subject: [PATCH 1/4] Add wb45n to uboot

Adds platform support for Laird WB45n. Add board definition and setup hardware.
---
 arch/arm/cpu/arm926ejs/at91/at91cap9_devices.c |   20 +-
 board/atmel/at91sam9x5ek/at91sam9x5ek.c        |    9 +-
 board/atmel/at91sam9x5ek/config.mk             |    2 +-
 boards.cfg                                     |    1 +
 include/configs/wb45n.h                        |  289 ++++++++++++++++++++++++
 5 files changed, 308 insertions(+), 13 deletions(-)
 create mode 100644 include/configs/wb45n.h

diff --git a/arch/arm/cpu/arm926ejs/at91/at91cap9_devices.c b/arch/arm/cpu/arm926ejs/at91/at91cap9_devices.c
index db2ecb8..564b32e 100644
--- a/arch/arm/cpu/arm926ejs/at91/at91cap9_devices.c
+++ b/arch/arm/cpu/arm926ejs/at91/at91cap9_devices.c
@@ -167,16 +167,16 @@ void at91_spi1_hw_init(unsigned long cs_mask)
 #ifdef CONFIG_MACB
 void at91_macb_hw_init(void)
 {
-	at91_set_a_periph(AT91_PIO_PORTB, 21, 0);	/* ETXCK_EREFCK */
-	at91_set_a_periph(AT91_PIO_PORTB, 22, 0);	/* ERXDV */
-	at91_set_a_periph(AT91_PIO_PORTB, 25, 0);	/* ERX0 */
-	at91_set_a_periph(AT91_PIO_PORTB, 26, 0);	/* ERX1 */
-	at91_set_a_periph(AT91_PIO_PORTB, 27, 0);	/* ERXER */
-	at91_set_a_periph(AT91_PIO_PORTB, 28, 0);	/* ETXEN */
-	at91_set_a_periph(AT91_PIO_PORTB, 23, 0);	/* ETX0 */
-	at91_set_a_periph(AT91_PIO_PORTB, 24, 0);	/* ETX1 */
-	at91_set_a_periph(AT91_PIO_PORTB, 30, 0);	/* EMDIO */
-	at91_set_a_periph(AT91_PIO_PORTB, 29, 0);	/* EMDC */
+	at91_set_a_periph(AT91_PIO_PORTB, 4, 0);	/* ETXCK_EREFCK */
+	at91_set_a_periph(AT91_PIO_PORTB, 3, 0);	/* ERXDV */
+	at91_set_a_periph(AT91_PIO_PORTB, 0, 0);	/* ERX0 */
+	at91_set_a_periph(AT91_PIO_PORTB, 1, 0);	/* ERX1 */
+	at91_set_a_periph(AT91_PIO_PORTB, 2, 0);	/* ERXER */
+	at91_set_a_periph(AT91_PIO_PORTB, 7, 0);	/* ETXEN */
+	at91_set_a_periph(AT91_PIO_PORTB, 9, 0);	/* ETX0 */
+	at91_set_a_periph(AT91_PIO_PORTB, 10, 0);	/* ETX1 */
+	at91_set_a_periph(AT91_PIO_PORTB, 5, 0);	/* EMDIO */
+	at91_set_a_periph(AT91_PIO_PORTB, 6, 0);	/* EMDC */
 
 #ifndef CONFIG_RMII
 	at91_set_b_periph(AT91_PIO_PORTC, 25, 0);	/* ECRS */
diff --git a/board/atmel/at91sam9x5ek/at91sam9x5ek.c b/board/atmel/at91sam9x5ek/at91sam9x5ek.c
index edb0886..3a81335 100644
--- a/board/atmel/at91sam9x5ek/at91sam9x5ek.c
+++ b/board/atmel/at91sam9x5ek/at91sam9x5ek.c
@@ -61,11 +61,12 @@ static void at91sam9x5ek_nand_hw_init(void)
 	/* Enable CS3 */
 	csa = readl(&matrix->ebicsa);
 	csa |= AT91_MATRIX_EBI_CS3A_SMC_SMARTMEDIA;
-	/* NAND flash on D16 */
-	csa |= AT91_MATRIX_NFD0_ON_D16;
+	/* NAND flash on D0 */
+	csa &= ~AT91_MATRIX_NFD0_ON_D16;
 
 	/* Configure IO drive */
 	csa &= ~AT91_MATRIX_EBI_EBI_IOSR_NORMAL;
+	csa &= ~AT91_MATRIX_MP_ON;
 
 	writel(csa, &matrix->ebicsa);
 
@@ -94,11 +95,14 @@ static void at91sam9x5ek_nand_hw_init(void)
 	at91_set_gpio_input(CONFIG_SYS_NAND_READY_PIN, 1);
 	/* Enable NandFlash */
 	at91_set_gpio_output(CONFIG_SYS_NAND_ENABLE_PIN, 1);
+	/* Disable Flash Write Protect Line */
+	at91_set_gpio_output(AT91_PIN_PD10, 1);
 
 	at91_set_a_periph(AT91_PIO_PORTD, 0, 1);	/* NAND OE */
 	at91_set_a_periph(AT91_PIO_PORTD, 1, 1);	/* NAND WE */
 	at91_set_a_periph(AT91_PIO_PORTD, 2, 1);	/* NAND ALE */
 	at91_set_a_periph(AT91_PIO_PORTD, 3, 1);	/* NAND CLE */
+/*
 	at91_set_a_periph(AT91_PIO_PORTD, 6, 1);
 	at91_set_a_periph(AT91_PIO_PORTD, 7, 1);
 	at91_set_a_periph(AT91_PIO_PORTD, 8, 1);
@@ -107,6 +111,7 @@ static void at91sam9x5ek_nand_hw_init(void)
 	at91_set_a_periph(AT91_PIO_PORTD, 11, 1);
 	at91_set_a_periph(AT91_PIO_PORTD, 12, 1);
 	at91_set_a_periph(AT91_PIO_PORTD, 13, 1);
+*/
 }
 #endif
 
diff --git a/board/atmel/at91sam9x5ek/config.mk b/board/atmel/at91sam9x5ek/config.mk
index 6589a12..e554a45 100644
--- a/board/atmel/at91sam9x5ek/config.mk
+++ b/board/atmel/at91sam9x5ek/config.mk
@@ -1 +1 @@
-CONFIG_SYS_TEXT_BASE = 0x26f00000
+CONFIG_SYS_TEXT_BASE = 0x23f00000
diff --git a/boards.cfg b/boards.cfg
index 35f38f3..cad7566 100644
--- a/boards.cfg
+++ b/boards.cfg
@@ -96,6 +96,7 @@ at91sam9g20ek_nandflash      arm         arm926ejs   at91sam9260ek       atmel
 at91sam9m10g45ek_nandflash   arm         arm926ejs   at91sam9m10g45ek    atmel          at91        at91sam9m10g45ek:AT91SAM9M10G45,SYS_USE_NANDFLASH
 at91sam9rlek_dataflash       arm         arm926ejs   at91sam9rlek        atmel          at91        at91sam9rlek:AT91SAM9RL,SYS_USE_DATAFLASH
 at91sam9rlek_nandflash       arm         arm926ejs   at91sam9rlek        atmel          at91        at91sam9rlek:AT91SAM9RL,SYS_USE_NANDFLASH
+wb45n                        arm         arm926ejs   at91sam9x5ek        atmel          at91        wb45n:AT91SAM9X5,SYS_USE_NANDFLASH
 at91sam9x5ek_nandflash       arm         arm926ejs   at91sam9x5ek        atmel          at91        at91sam9x5ek:AT91SAM9X5,SYS_USE_NANDFLASH
 at91sam9x5ek_spiflash        arm         arm926ejs   at91sam9x5ek        atmel          at91        at91sam9x5ek:AT91SAM9X5,SYS_USE_SPIFLASH
 at91sam9x5ek_mmc             arm         arm926ejs   at91sam9x5ek        atmel          at91        at91sam9x5ek:AT91SAM9X5,SYS_USE_MMC
diff --git a/include/configs/wb45n.h b/include/configs/wb45n.h
new file mode 100644
index 0000000..d956197
--- /dev/null
+++ b/include/configs/wb45n.h
@@ -0,0 +1,289 @@
+/*
+ * Copyright (C) 2012 Atmel Corporation
+ *
+ * Configuation settings for the AT91SAM9X5EK board.
+ *
+ * See file CREDITS for list of people who contributed to this
+ * project.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+ * MA 02111-1307 USA
+ */
+
+#ifndef __CONFIG_H__
+#define __CONFIG_H__
+/*
+ * based on at91sam9x5ek_nandflash_config 
+ * for the wb45nbt
+ *
+ * tabs removed from whitespace, values column at 41
+ */
+//#define CONFIG_DYNAMIC_BOOTARGS
+
+
+
+#include <asm/hardware.h>
+
+/* ARM asynchronous clock */
+#define CONFIG_SYS_AT91_SLOW_CLOCK      32768
+#define CONFIG_SYS_AT91_MAIN_CLOCK      12000000   /* 12 MHz crystal */
+#define CONFIG_SYS_HZ                   1000
+
+#define CONFIG_AT91SAM9X5EK
+#define CONFIG_AT91FAMILY
+
+#define CONFIG_CMDLINE_TAG              /* enable passing of ATAGs */
+#define CONFIG_SETUP_MEMORY_TAGS
+#define CONFIG_INITRD_TAG
+#define CONFIG_SKIP_LOWLEVEL_INIT
+#define CONFIG_BOARD_EARLY_INIT_F
+#define CONFIG_DISPLAY_CPUINFO
+
+/* #define CONFIG_OF_LIBFDT */
+
+/* general purpose I/O */
+#define CONFIG_ATMEL_LEGACY      /* required until (g)pio is fixed */
+#define CONFIG_AT91_GPIO
+//-?-#define CONFIG_AT91_GPIO_PULLUP  1  /* keep pullups on peripheral pins */
+
+/* serial console */
+#define CONFIG_ATMEL_USART
+#define CONFIG_USART_BASE               ATMEL_BASE_DBGU
+#define CONFIG_USART_ID                 ATMEL_ID_SYS
+
+/* LCD */
+#undef CONFIG_LCD
+//#define LCD_BPP         LCD_COLOR16
+//#define LCD_OUTPUT_BPP      24
+//#define CONFIG_LCD_LOGO
+//#undef LCD_TEST_PATTERN
+//#define CONFIG_LCD_INFO
+//#define CONFIG_LCD_INFO_BELOW_LOGO
+//#define CONFIG_SYS_WHITE_ON_BLACK
+//#define CONFIG_ATMEL_HLCD
+//#define CONFIG_ATMEL_LCD_RGB565
+//#define CONFIG_SYS_CONSOLE_IS_IN_ENV
+
+/* LED */
+#define CONFIG_AT91_LED
+//#define CONFIG_RED_LED                  AT91_PIN_PA25  /* led0 */
+//#define CONFIG_GREEN_LED                AT91_PIN_PA26  /* led1 */
+
+/*
+ * BOOTP options
+ */
+#define CONFIG_BOOTP_BOOTFILESIZE
+#define CONFIG_BOOTP_BOOTPATH
+#define CONFIG_BOOTP_GATEWAY
+#define CONFIG_BOOTP_HOSTNAME
+
+/*
+ * Command line configuration.
+ */
+#include <config_cmd_default.h>
+#undef CONFIG_CMD_FPGA
+#undef CONFIG_CMD_IMI
+#undef CONFIG_CMD_IMLS
+#undef CONFIG_CMD_LOADS
+
+#define CONFIG_CMD_PING
+#define CONFIG_CMD_DHCP
+#define CONFIG_CMD_NAND
+#undef CONFIG_CMD_SF
+#define CONFIG_CMD_MMC
+#define CONFIG_CMD_MII
+
+/* SDRAM */
+#define CONFIG_NR_DRAM_BANKS            1
+#define CONFIG_SYS_SDRAM_BASE           0x20000000
+#define CONFIG_SYS_SDRAM_SIZE           0x04000000   /* 64 MB */
+
+#define CONFIG_SYS_INIT_SP_ADDR \
+       (CONFIG_SYS_SDRAM_BASE + 4 * 1024 - GENERATED_GBL_DATA_SIZE)
+
+/* DataFlash */
+#ifdef CONFIG_CMD_SF
+#define CONFIG_ATMEL_SPI
+#undef CONFIG_SPI_FLASH
+#undef CONFIG_SPI_FLASH_ATMEL
+#define CONFIG_SF_DEFAULT_SPEED         30000000
+#endif
+
+/* no NOR flash */
+#define CONFIG_SYS_NO_FLASH
+
+/* NAND flash */
+#ifdef CONFIG_CMD_NAND
+#define CONFIG_NAND_ATMEL
+#define CONFIG_SYS_MAX_NAND_DEVICE      1
+#define CONFIG_SYS_NAND_BASE            0x40000000
+#define CONFIG_SYS_NAND_DBW_8           1
+#define CONFIG_SYS_NAND_MASK_ALE        (1 << 21)  /* ALE is AD21 */
+#define CONFIG_SYS_NAND_MASK_CLE        (1 << 22)  /* CLE is AD22 */
+#define CONFIG_SYS_NAND_ENABLE_PIN      AT91_PIN_PD4
+#define CONFIG_SYS_NAND_READY_PIN       AT91_PIN_PD5
+#define CONFIG_CMD_NAND_TRIMFFS         /* must trim 0xFFs for ubi rootfs */
+
+/* PMECC & PMERRLOC */
+#define CONFIG_ATMEL_NAND_HWECC         1
+#define CONFIG_ATMEL_NAND_HW_PMECC      1
+#define CONFIG_PMECC_CAP                4
+#define CONFIG_PMECC_SECTOR_SIZE        512
+#define CONFIG_PMECC_INDEX_TABLE_OFFSET 0x8000
+
+#define CONFIG_MTD_DEVICE
+#define CONFIG_CMD_MTDPARTS
+#define CONFIG_MTD_PARTITIONS
+#define CONFIG_RBTREE
+#define CONFIG_LZO
+#define CONFIG_CMD_UBI
+#define CONFIG_CMD_UBIFS
+#endif
+
+/* MMC */
+#ifdef CONFIG_CMD_MMC
+#define CONFIG_MMC
+#define CONFIG_CMD_FAT
+#define CONFIG_GENERIC_MMC
+#define CONFIG_GENERIC_ATMEL_MCI
+#define CONFIG_DOS_PARTITION
+#endif
+
+/* Ethernet */
+#define CONFIG_MACB
+#define CONFIG_RMII
+#define CONFIG_NET_RETRY_COUNT          20
+#define CONFIG_MACB_SEARCH_PHY
+#define CONFIG_ETHADDR                  00:17:23:00:00:00
+
+#define CONFIG_ENV_OVERWRITE     1
+//#define CONFIG_OVERWRITE_ETHADDR_ONCE   1
+
+
+
+
+#define CONFIG_SYS_LOAD_ADDR            0x22000000   /* load address */
+
+#define CONFIG_SYS_MEMTEST_START        CONFIG_SYS_SDRAM_BASE
+#define CONFIG_SYS_MEMTEST_END          0x23e00000
+
+
+#ifdef CONFIG_SYS_USE_NANDFLASH
+#define CONFIG_ENV_IS_IN_NAND
+#define CONFIG_ENV_OFFSET               0xa0000
+#define CONFIG_ENV_OFFSET_REDUND        0xc0000
+/* #define CONFIG_ENV_RANGE 0x40000 */
+#define CONFIG_ENV_SIZE                 0x20000      /* 1 sector = 128 kB */
+
+#define CONFIG_BOOTCOMMAND              "nand read 0x22000000 0x000e0000 0x280000; "\
+                                        "run _mtd; bootm"
+
+#define CONFIG_BOOTARGS                 "console=ttyS0,115200 loglevel=4 " \
+                                        "rw noinitrd mem=64M " \
+                                        "rootfstype=ubifs root=ubi0:rootfs ubi.mtd=6"
+/*
+#define CONFIG_BOOTARGS                 "mem=64M console=ttyS0,115200 " \
+                                        "mtdparts=atmel_nand:" \
+                                        "8M(bootstrap/uboot/kernel)ro,-(rootfs) " \
+                                        "root=/dev/mtdblock1 rw " \
+                                        "rootfstype=ubifs ubi.mtd=1 root=ubi0:rootfs"
+*/
+                                        
+#else /* not using nandflash */
+#ifdef CONFIG_SYS_USE_SPIFLASH
+   /* bootstrap + u-boot + env + linux in spi flash */
+#define CONFIG_ENV_IS_IN_SPI_FLASH
+#define CONFIG_ENV_OFFSET   0x5000
+#define CONFIG_ENV_SIZE      0x3000
+#define CONFIG_ENV_SECT_SIZE   0x1000
+#define CONFIG_ENV_SPI_MAX_HZ   30000000
+#define CONFIG_BOOTCOMMAND   "sf probe 0; " \
+            "sf read 0x22000000 0x100000 0x300000; " \
+            "bootm 0x22000000"
+#endif
+#endif
+
+/* Misc Environment Vars */
+#define CONFIG_BOOTDELAY                1
+#define CONFIG_BAUDRATE                 115200
+#define CONFIG_SYS_PROMPT               "U-Boot> "
+#define CONFIG_SYS_CBSIZE               256
+#define CONFIG_SYS_MAXARGS              16
+#define CONFIG_SYS_PBSIZE               \
+   (CONFIG_SYS_CBSIZE + sizeof(CONFIG_SYS_PROMPT) + 16)
+#define CONFIG_SYS_LONGHELP
+#define CONFIG_CMDLINE_EDITING
+#define CONFIG_AUTO_COMPLETE
+#define CONFIG_SYS_HUSH_PARSER
+
+/* Size of malloc() pool */
+#define CONFIG_SYS_MALLOC_LEN           (512 * 1024 + 0x1000)
+
+#define MTDIDS_DEFAULT                  "nand0=atmel_nand"
+#define MTDPARTS_DEFAULT                "mtdparts=atmel_nand:" \
+                                        "128K(at91bs)," \
+                                        "512K(u-boot)," \
+                                        "128K(u-boot-env)," \
+                                        "128K(redund-env)," \
+                                        "2560K(kernel-a)," \
+                                        "2560K(kernel-b)," \
+                                        "38912K(rootfs-a)," \
+                                        "38912K(rootfs-b)," \
+                                        "46208K(user)," \
+                                        "512K(logs)"
+
+#ifdef CONFIG_USE_IRQ
+#error CONFIG_USE_IRQ not supported
+#endif
+
+#ifdef CONFIG_DYNAMIC_BOOTARGS
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND              "run _kernel-a _rootfs-a _wbt _mtd _xargs && bootm"
+#undef CONFIG_BOOTARGS
+#define CONFIG_BOOTARGS                 "console=ttyS0,115200 loglevel=4 rw mem=64M"
+/*
+ * Setup the wb45n platform with a 'dynamic-bootargs' boot environment.
+ * The prefixed variables are used to support this feature.
+ */
+#define CONFIG_EXTRA_ENV_SETTINGS       \
+        "-msg=set -msg\; echo\; echo Initializing dynamic-bootargs support.\0" \
+        "-set=set -set\; set bootdelay 1\0" \
+        "-wb=run -set -msg\0" \
+        "_kernel-a=nand read 0x22000000 0x000e0000 0x280000; ki=a\0" \
+        "_kernel-b=nand read 0x22000000 0x00360000 0x280000; ki=b\0" \
+        "_mtd=test -n ${mtdparts} || mtdparts default\0" \
+        "_rootfs-a=fs=rootfstype=ubifs\\\\ root=ubi0:rootfs\\\\ ubi.mtd=6; ri=a\0" \
+        "_rootfs-b=fs=rootfstype=ubifs\\\\ root=ubi0:rootfs\\\\ ubi.mtd=7; ri=b\0" \
+        "_wbt=wbti=${wbtype}=${ki}/${ri}; if ${-wb}; then set -wb; save; fi\0" \
+        "_xargs=set bootargs ${bootargs} ${wbti} ${fs} ${mtdparts}\0" \
+        "autoload=no\0" \
+        "bootdelay=0\0" \
+        "wbtype=wb45nbt\0" \
+        "\0"
+#else
+/*
+ * Setup (default) minimal support for platform-init and fw_update/select.
+ */
+#define CONFIG_EXTRA_ENV_SETTINGS       \
+        "_mtd=mtdparts default; set bootargs ${bootargs} ${wbti} ${mtdparts}\0" \
+        "autoload=no\0" \
+        "\0"
+#endif
+
+/* testing */
+//#define CONFIG_HW_WATCHDOG              1
+//#define CONFIG_AT91SAM9_WATCHDOG        1
+#endif
+
-- 
1.7.9.5

